### ISSUE TITLE
[BUG] [alerting_exception] analyzer [analyzer_keyword] has not been configured in mappings

### ISSUE THREAD

**Issue by paasi6666** [opensearch-project/alerting#961]
**What is the bug?**
When defining a new monitor (under alerting) and selecting the type 'Per document monitor', the monitor saves with following error:
![alert_monitoring](https://github.com/opensearch-project/security-analytics/assets/26493612/f8fe80fe-0768-4fea-8428-cc70ac803146)

Furthermore, when "testing" the query it times out:
![alert_monitoring_timeout](https://github.com/opensearch-project/security-analytics/assets/26493612/09140c2a-9cfd-46fc-a880-4f3e694ed799)

The query looks like this:
![image](https://github.com/opensearch-project/security-analytics/assets/26493612/6a536d2c-1158-47a6-8e42-edcaadcc00c2)

and as shown below, the query works when discovering:
![image](https://github.com/opensearch-project/security-analytics/assets/26493612/b3824757-2674-4b9a-9f6b-043c5b91325c)

**How can one reproduce the bug?**
1. Go to Alerting>Monitors>Create monitor
2. Select 'Per document monitor', select any index and choose a query
3. Go to Preview query and performance and wait..
4. Try to save the monitor

**What is the expected behavior?**
I don't know, hence it never worked for me.

**What is your host/environment?**
- OS: Centos7
- Opensearch Version: 2.7.0
- Opensearch-Dashboards Version: 2.7.0
-

**NOTE**
We are ingesting the logs using graylog.

**Comment by AWSHurneyt** [comment]
Transferred this issue to the Alerting plugin folder as alerting owns development and maintenance of Document level monitors.

**Comment by lezzago** [comment]
@paasi6666 Could you please share your index mapping for us to reproduce the issue?

**Comment by paasi6666** [comment]
Sure. Note, that the mapping is generated by graylog.

```
{
  "rpz_0": {
    "mappings": {
      "dynamic_templates": [
        {
          "internal_fields": {
            "match": "gl2_*",
            "match_mapping_type": "string",
            "mapping": {
              "type": "keyword"
            }
          }
        },
        {
          "store_generic": {
            "match_mapping_type": "string",
            "mapping": {
              "type": "keyword"
            }
          }
        }
      ],
      "properties": {
        "@metadata_beat": {
          "type": "keyword"
        },
        "@metadata_type": {
          "type": "keyword"
        },
        "@metadata_version": {
          "type": "keyword"
        },
        "@timestamp": {
          "type": "date"
        },
        "agent_ephemeral_id": {
          "type": "keyword"
        },
        "agent_name": {
          "type": "keyword"
        },
        "beats_type": {
          "type": "keyword"
        },
        "client_id": {
          "type": "keyword"
        },
        "event_action": {
          "type": "keyword"
        },
        "full_message": {
          "type": "text",
          "analyzer": "standard"
        },
        "gl2_accounted_message_size": {
          "type": "long"
        },
        "gl2_message_id": {
          "type": "keyword"
        },
        "gl2_processing_error": {
          "type": "keyword"
        },
        "gl2_processing_timestamp": {
          "type": "date",
          "format": "uuuu-MM-dd HH:mm:ss.SSS"
        },
        "gl2_receive_timestamp": {
          "type": "date",
          "format": "uuuu-MM-dd HH:mm:ss.SSS"
        },
        "gl2_remote_ip": {
          "type": "keyword"
        },
        "gl2_remote_port": {
          "type": "long"
        },
        "gl2_source_input": {
          "type": "keyword"
        },
        "gl2_source_node": {
          "type": "keyword"
        },
        "host_name": {
          "type": "keyword"
        },
        "hostname": {
          "type": "keyword"
        },
        "log_file_path": {
          "type": "keyword"
        },
        "log_offset": {
          "type": "long"
        },
        "loglevel": {
          "type": "keyword"
        },
        "message": {
          "type": "text",
          "analyzer": "standard"
        },
        "query_action": {
          "type": "keyword"
        },
        "query_class": {
          "type": "keyword"
        },
        "query_name": {
          "type": "keyword"
        },
        "query_type": {
          "type": "keyword"
        },
        "rpz_category": {
          "type": "keyword"
        },
        "rpz_message": {
          "type": "keyword"
        },
        "rpz_zone": {
          "type": "keyword"
        },
        "source": {
          "type": "text",
          "analyzer": "analyzer_keyword",
          "fielddata": true
        },
        "source_ip": {
          "type": "keyword"
        },
        "source_port": {
          "type": "keyword"
        },
        "streams": {
          "type": "keyword"
        },
        "timestamp": {
          "type": "date",
          "format": "uuuu-MM-dd HH:mm:ss.SSS"
        },
        "url_domain": {
          "type": "keyword"
        },
        "url_short": {
          "type": "keyword"
        }
      }
    }
  }
}
```

**Comment by paasi6666** [comment]
@lezzago The settings of the index look like follows:

```
GET rpz_0/_settings
{
  "rpz_0": {
    "settings": {
      "index": {
        "number_of_shards": "4",
        "provided_name": "rpz_0",
        "creation_date": "1649938793819",
        "analysis": {
          "analyzer": {
            "analyzer_keyword": {
              "filter": "lowercase",
              "tokenizer": "keyword"
            }
          }
        },
        "number_of_replicas": "0",
        "uuid": "e8NRlQCHQfau984C3QGMPQ",
        "version": {
          "created": "7100299",
          "upgraded": "136287827"
        }
      }
    }
  }
}
```

**Comment by pawelw1** [comment]
I've tested this issue with an example from Elastic https://www.elastic.co/guide/en/elasticsearch/reference/current/analyzer.html

The result was exactly the same 
```
opensearch-node1_2.6.0   | org.opensearch.alerting.util.AlertingException: analyzer [my_analyzer] has not been configured in mappings
opensearch-node1_2.6.0   |      at org.opensearch.alerting.util.AlertingException$Companion.wrap(AlertingException.kt:70) ~[opensearch-alerting-2.6.0.0.jar:2.6.0.0]
opensearch-node1_2.6.0   |      at org.opensearch.alerting.util.DocLevelMonitorQueries.updateQueryIndexMappings(DocLevelMonitorQueries.kt:369) ~[opensearch-alerting-2.6.0.0.jar:2.6.0.0]
opensearch-node1_2.6.0   |      at org.opensearch.alerting.util.DocLevelMonitorQueries.access$updateQueryIndexMappings(DocLevelMonitorQueries.kt:45) ~[opensearch-alerting-2.6.0.0.jar:2.6.0.0]
opensearch-node1_2.6.0   |      at org.opensearch.alerting.util.DocLevelMonitorQueries$updateQueryIndexMappings$1.invokeSuspend(DocLevelMonitorQueries.kt) ~[opensearch-alerting-2.6.0.0.jar:2.6.0.0]
opensearch-node1_2.6.0   |      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) [kotlin-stdlib-1.6.10.jar:1.6.10-release-923(1.6.10)]
opensearch-node1_2.6.0   |      at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:285) [kotlinx-coroutines-core-1.1.1.jar:?]
opensearch-node1_2.6.0   |      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594) [kotlinx-coroutines-core-1.1.1.jar:?]
opensearch-node1_2.6.0   |      at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60) [kotlinx-coroutines-core-1.1.1.jar:?]
opensearch-node1_2.6.0   |      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742) [kotlinx-coroutines-core-1.1.1.jar:?]
opensearch-node1_2.6.0   | Caused by: java.lang.Exception: java.lang.IllegalArgumentException: analyzer [my_analyzer] has not been configured in mappings
opensearch-node1_2.6.0   |      ... 9 more
opensearch-node1_2.6.0   | [2023-06-21T15:04:14,379][ERROR][o.o.a.u.AlertingException] [opensearch-node1] Alerting error: AlertingException[analyzer [my_analyzer] has not been configured in mappings]; nested: Exception[java.lang.IllegalArgumentException: analyzer [my_analyzer] has not been configured in mappings];
```
I've tested that analyzer with examples from the link and it works with no issues but for some reason Alerting plugin can't see the definition of that analyzer when creating a Monitor. 
Looking at the logs, it looks like the Analyzer plugin trigger errors when the index is selected in the Monitor and before hitting the create button. I assume there is a validation running before the Monitor is created.

**Comment by petardz** [comment]
This is likely because Alerting plugin doesn't copy analyzer def from source index settings to queryIndex settings.

**Comment by eirsep** [comment]
Analyzer updates are static config changes to an index. they would require closing an index > apply analyzer setting change > re-open index.

Closing alerting query index is not possible as all monitors share query index and monitors are running in parallel.
We cannot support this in current architecture

**Comment by mvanderlee** [comment]
@eirsep So any index with a custom analyzer or normalizer simply can not have a monitor configured on it? And this is acceptable to the OpenSearch team?!
Since this is tied to security detectors as well, this should be a considered a major bug, not to be closed lightly!

https://github.com/opensearch-project/security-analytics/issues/697

**Comment by dblock** [comment]
Looks like this was reopened, thanks for your feedback @mvanderlee. 

[Catch All Triage w/ [1](https://github.com/dblock), [2](https://github.com/krisfreedain), [3](https://github.com/andrross)]

**Comment by christiand93** [comment]
Hi all,
We are facing this issue as well now (OpenSearch v2.19.5). Is there any plan to mitigate/fix this problem?

**Comment by thecodingshrimp** [comment]
**The "won't fix" reasoning does not apply to the strip-attributes approach — a low-risk fix is available**

I've traced this bug through the source code of both the alerting and security-analytics repos to understand the full picture. Here is what I found.

**The two distinct fix strategies, and why only one was evaluated**

The original closure cited that applying static analyzer settings to the query index requires closing it, which is impossible while monitors run in parallel. That is correct — but it describes only one of two possible fixes (Strategy B: copy `settings.analysis.*` from the source index into the query index). A second approach was never evaluated.

**Strategy A — strip analysis attributes (recommended, no close/open required)**

In `DocLevelMonitorQueries.kt`, the `leafNodeProcessor` closure (~line 329) does:
```kotlin
val newProps = props.toMutableMap()
```
This is a verbatim copy of every source field attribute, including `analyzer`, `normalizer`, `similarity`, `search_analyzer`, and `search_quote_analyzer`. These attributes reference custom analysis objects defined in the source index's `settings.analysis.*` block, which is never replicated to the query index (whose settings contain only `{ "index": { "mapping": { "total_fields": { "limit": 10000 } }, "hidden": true } }`). OpenSearch rejects the resulting `PutMappingRequest` because the referenced analyzer/normalizer does not exist in the query index's settings.

The fix: strip these attributes from `newProps` before the `PutMappingRequest` is sent:

```kotlin
val ANALYSIS_ATTRIBUTES = setOf("analyzer", "normalizer", "similarity", "search_analyzer", "search_quote_analyzer")
val newProps = props.toMutableMap().apply { keys.removeAll(ANALYSIS_ATTRIBUTES) }
```

This change:
- Requires no index close/open cycle — it modifies the in-memory `PutMappingRequest` payload only
- Does not conflict with the shared-query-index architecture
- Is safe and lossless for the query index's purpose: the query index stores Percolator query documents for doc-level matching, not user-facing search data. Fields falling back to the default analyzer on the query index does not affect the correctness of doc-level monitor matching.

**Why Strategy B is actually blocked (and Strategy A is not)**

Strategy B (copying `analysis.*` settings at query index creation time) is blocked — but not for the close/open reason. The real blocker is that the query index is shared across all monitors, which may span source indices with conflicting custom analyzer definitions of the same name. Merging conflicting `analysis.*` blocks from different source indices into one shared index is not safe. Strategy A avoids this entirely.

**Impact**

This bug causes active production failures in opensearch-project/security-analytics (issues #697 and #1798). The failure surface is broader than `normalizer` alone: any source index field carrying `analyzer`, `search_analyzer`, `search_quote_analyzer`, `normalizer`, or `similarity` attributes that reference custom analysis objects will trigger the same `PutMappingRequest` rejection. This blocks detector creation on any real-world index that uses custom analyzers or normalizers — a common pattern in production clusters.

The bug remains reproducible on OpenSearch v2.19.5 (confirmed in a recent comment on this issue).

**Ask**

Please consider reopening specifically for the strip-attributes fix (Strategy A). The architectural objection that led to the initial closure applies to Strategy B, not Strategy A. These are two separate solutions and only one of them was evaluated before the issue was closed.

credit: claude
